### DISCUSSION TITLE
translate is really quite slow

### DISCUSSION THREAD

**Discussion by wnavarre** [geopandas/geopandas#2930] [Ideas]
Translating a series of points is surprisingly slow. 

In my project I've wound up implementing `my_translate` (similar to below) which runs much more quickly. 

I wonder if geopandas could do an optimization like this automatically when applicable?

```py
from shapely import Point, LineString, Polygon
import geopandas as gpd
import random

random.seed(30)

SERIES = gpd.GeoSeries([Point(random.random(), random.random()) for _ in range(100 * 1000)])
TRANSLATE_ARGS = 1.2, 3.7
SHAPE = gpd.GeoSeries([Polygon([(0., 0.), (0., .4), (1., 1.), (1., 0.), (0., 0.)])] * (100 * 1000))

def something_harder(): return SHAPE.covers(SERIES)

def geopandas_translate(): return SERIES.translate(*TRANSLATE_ARGS)

def my_translate():
    delta_x, delta_y = TRANSLATE_ARGS
    new_x = SERIES.x + delta_x
    new_y = SERIES.y + delta_y
    return gpd.GeoSeries.from_xy(new_x, new_y)

def check():
    assert (my_translate() == geopandas_translate()).all()
    return "OK"
```

Using `%timeit`, `my_translate()` is much faster than `geopandas_translate()`. (50 times faster!!!)

Even more surprisingly, `geopandas_translate()` is also much slower than `something_harder()`!!!

```py
In [1]: from translate_test import *

In [2]: %timeit my_translate()
38.2 ms ± 66.7 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

In [3]: %timeit geopandas_translate()
2.05 s ± 6.84 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [4]: %timeit something_harder()
264 ms ± 4.18 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [5]: check()
Out[5]: 'OK'
```

I am running geopandas 0.12.2 (to be honest I'm not sure it's possible to upgrade beyond that with `conda` at the moment?). 

I happened to notice a similar problems with `rotate` and had to implement my own version similar to what I've done here with `my_translate`.

**Comment by martinfleis** [comment]
Yeah, our implementation is super old and pre-dates all the work on vectorized geometry engines (pygeos, shapely 2.0). There's clearly a space to improve the performance there.

There's still a for loop in there which can be replaced by `shapely.transform` for `translate`. Not sure about other affine transformations like `rotate`...

https://github.com/geopandas/geopandas/blob/3a45af899717de473da023070fab1e4719de7890/geopandas/_vectorized.py#L453-L461

**Reply by wnavarre** [reply]
Sounds good. 

Right now I think it's much slower than developers will expect; it took me quite a while to realize `translate()` of all things was making my program run slow. 

You sort of would imagine `translate()` would be at least as fast as some other operations like `covers`... 

Wonder if there would be some (documented) way to help developers identify when they are using an unexpectedly slow function...
